### ISSUE TITLE
[Bug Report] v0 drop of Llama3.1-70B on LoudBox hangs on latest firmware [80.13.1.0] when running full dataset evaluation

### ISSUE THREAD

**Issue by milank94** [tenstorrent/tt-metal#15090]
**Describe the bug**
When running the latest drop (v0) of Llama3.1-70B on LoudBox with firmware [80.13.1.0](https://github.com/tenstorrent/tt-firmware/blob/main/fw_pack-80.13.1.0.fwbundle) we experience a hang (presumably di/dt). Running on firmware with downgraded clock of 900MHz, there is no hang. Validated on [80.13.1.1](https://github.com/tenstorrent/tt-firmware/blob/main/experiments/fw_pack-80.13.1.1.tar.gz) and [80.10.1.0](https://github.com/tenstorrent/tt-firmware/blob/02b4b6ed49b6ea2fb9a8664e99d4fed25e443bd6/experiments/fw_pack-80.10.1.0.fwbundle).

**To Reproduce**

1. Install tt-metal + vllm for the latest drop of Llama3.1-70B on LoudBox following the instructions from: <https://github.com/tenstorrent/vllm/tree/dev/tt_metal> (this includes downloading the model weights)
2. With the vllm environment from step 1. active, from your home directory, run:
```bash
git clone https://github.com/tenstorrent/tt-inference-server.git
cd tt-inference-server/vllm-tt-metal-llama3-70b/src
pip install pyjwt==2.7.0
export JWT_SECRET=password
WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml python run_vllm_api_server.py
```
3. In a new terminal, activate the environment from step 1., and after step 2. has successfully deployed, run:
```bash
export JWT_SECRET=password
export CACHE_ROOT=$PWD
python example_openai_client_alpaca_eval.py
```

If running on 80.13.1.0 (1000MHz), the model will eventually hang. This usually occurs at a random point.

**Please complete the following environment information:**
- OS: Ubuntu 20.04
- Version of software (tt-metal): [3859041](https://github.com/tenstorrent/tt-metal/tree/385904186f81fed15d5c87c162221d4f34387164)
- Version of software (vllm): [384f179](https://github.com/tenstorrent/vllm/tree/384f1790c3be16e1d1b10de07252be2e66d00935)
- TT-KMD: 1.29
- Firmware: 
    - [80.13.1.0](https://github.com/tenstorrent/tt-firmware/blob/main/fw_pack-80.13.1.0.fwbundle) [hang]
    - [80.10.1.0](https://github.com/tenstorrent/tt-firmware/blob/02b4b6ed49b6ea2fb9a8664e99d4fed25e443bd6/experiments/fw_pack-80.10.1.0.fwbundle) [pass]
    - [80.13.1.1](https://github.com/tenstorrent/tt-firmware/blob/main/experiments/fw_pack-80.13.1.1.tar.gz) [pass]

**Additional context**
Tested this on a few systems, all experiencing the same behaviour. Going from 1000MHz -> 900MHz was able to stabilize.

**Comment by uaydonat** [comment]
@cglagovichTT can you post a table; for each matmul in the 70B, what are the dims, and which matmul type we are using (1D, 2D) and how many cores?

the feedback from @ttmtrajkovic is that only 1D matmuls with 64 cores still exhibit di/di issues

@milank94 I think we decided last friday that you guys will re-test llama on "good" systems with the latest fw release and let us know if you are seeing the hang?

**Comment by milank94** [comment]
@uaydonat, yes I retested Llama on a good system using the latest fw release (1GHz) and a patch version with 900MHz.

- [8.13.1.0](https://github.com/tenstorrent/tt-firmware/blob/main/fw_pack-80.13.1.0.fwbundle) (1GHz) - test failed after ~12 hours. Ran it twice, both times got a hang.
- [8.13.1.1](https://github.com/tenstorrent/tt-firmware/blob/main/experiments/fw_pack-80.13.1.1.tar.gz) (900MHz) -  test successfully ran for >48 hours.

If you need access to the system, I can help setup the test or let it run until it hangs. You just need to be on Austin COLO VPN + I need to add your SSH Key.

**Comment by cglagovichTT** [comment]
| Name | Mode | Matmul Shape | Cores | Type | Variant |
|------|------|--------------|--------|------|----------|
| QKV | decode | 32 x 8192 x 1280 | 40 | HiFi2 BF16 x BFP8 => BF16 | matmul 1D |
| Out | decode | 32 x 1024 x 8192 | 34 | HiFi2 BF16 x BF16 => BF16 | Fused AllGather/matmul 1D |
| FF1 | decode | 32 x 8192 x 3584 | 64 | LoFi BF16 x BFP4 => BF16 | dram-sharded |
| FF3 | decode | 32 x 8192 x 3584 | 64 | LoFi BF16 x BFP4 => BF16 | dram-sharded |
| FF2 | decode | 32 x 3584 x 8192 | 64 | HiFi2 BFP8 x BFP8 => BFP8 | dram-sharded |
| LM head | decode | 32 x 8192 x 16384 | 32 | HiFi2 BF16 x BFP8 => BF16 | matmul 1D |
| QKV | prefill | 1024 x 8192 x 1280 | 64 | HiFi2 BF16 x BFP8 => BF16 | matmul 2D |
| Out | prefill | 1024 x 8192 x 1024 | 64 | HiFi2 BF16 x BFP8 => BF16 | matmul 2D |
| FF1 | prefill | 1024 x 8192 x 3584 | 56 | LoFi BF16 x BFP4 => BF16 | matmul 2D |
| FF3 | prefill | 1024 x 8192 x 3584 | 56 | LoFi BF16 x BFP4 => BF16 | matmul 2D |
| FF2 | prefill | 1024 x 3584 x 8192 | 64 | HiFi2 BFP8 x BFP8 => BFP8 | matmul 2D |
| LM head | prefill | 1024 x 8192 x 16384 | 64 | HiFi2 BF16 x BFP8 => BF16 | matmul 2D |

Few notes:
- I believe @milank94 and @tstescoTT see hangs in decode and not prefill, so we can ignore the prefill MMs. The workaround I've given them in the past slowed down decode MMs and resolved the hangs.
- dram-sharded matmuls reports that they run on 64 cores but this isn't accurate. They only run on the 12 cores adjacent to DRAM.

**Comment by milank94** [comment]
@cglagovichTT how did we conclude it was from decode and not prefill? From a recent thread with Davor:

> Regarding TTFT, ie prefill:
> - team hasn’t been focusing on prefill , it can be optimized much more
> - In prefill we start hitting di / dt and need to slow down Matmuls

**Comment by cglagovichTT** [comment]
@milank94 @tstescoTT I recall that when we first started seeing this Llama hang on the new Loudboxes, Tom reported that the hang occurred during decode steps. In addition, the workaround I gave Tom which solved the hangs slowed down the decode matmuls. 

I could be mistaken if this hang is different from those decode hangs you reported in early October

**Comment by milank94** [comment]
@cglagovichTT got it. Those systems/cards seem to have a different issue that we need to debug here: https://github.com/tenstorrent/cloud/issues/3217. May or may not be di/dt related since 900MHz was failing as well. 

For this issue, we're just going to focus on running on systems we know can execute to completion on 900MHz, but fail on 1GHz to isolate this. 

I would suggest we separate the two issues we are seeing. We may need to rerun some tests/analysis from before.

**Comment by uaydonat** [comment]
@ttmtrajkovic based on Colman's MM configs above, do you think di/dt is plausible?

@cglagovichTT we will need to replicate this on our side and try some work-arounds (disable dram sharded matmuls, reduce core grid, etc?) to get this model working with some reasonable perf loss. And, also based on Milos' feedback (if nothing is sus) we might need to identify hanging matmul and create a unit test them. Good news, we can just run these tests in the background while still working on other stuff.  

@milank94 can you setup the systems for Colman. 

thanks guys!

**Comment by ttmtrajkovic** [comment]
@cglagovichTT, thanks for the breakdown.
Given that @milank94's testing is much longer, in a setting not covered by the Model team, I'd say that both options - decode or prefil, are possible to cause problems. 

@uaydonat,
Looking at the matmul sizes and assuming dram-sharded are not at 64 cores, based on what we've tested so far this is my comment: 
- 2D/hifi2 shouldn't fail with di/dt problems based on the tests we've done
- 2D/lofi at 56 cores could be a suspect, but a very low probability since we've generally seen 2D matmuls (even with lofi) fail rarely, almost not at all with latest FW

I think the next steps are to isolate the op that is failing and take it from there.

**Comment by cglagovichTT** [comment]
update: 

Milan is running the repro with a patch to slow down decode matmuls. We will check back later today to see if it is stable. https://github.com/tenstorrent/tt-inference-server/blob/tstesco/llama3-program-configs/vllm-tt-metal-llama3-70b/configs/model_config_slow.py

**Comment by uaydonat** [comment]
Some potential next steps we thought of:

1. Run a decode-only test to re-confirm the hang is in decode.
2. Create unit tests for each of the matmuls in decode, run them all in parallel on different machines (for faster analysis) to isolate which matmul causes the hangs.
3. If we still do not have a lead, bisect Colman's fixes to narrow down which work-around fixes the issue.
4. Or, run even smaller tests (attention, mlp, LN) to narrow down the hang.

**Comment by milank94** [comment]
> update:
> 
> Milan is running the repro with a patch to slow down decode matmuls. We will check back later today to see if it is stable. https://github.com/tenstorrent/tt-inference-server/blob/tstesco/llama3-program-configs/vllm-tt-metal-llama3-70b/configs/model_config_slow.py

Update: the system crashed/hung even with the slow config. Occurred just after 24 hours of testing. As a reminder, this system was running firmware with 1GHz + slow config of decode matmuls. When tested with 900MHz + fast config of decode matmuls it successful passed over 48 hours of testing.

**Comment by cglagovichTT** [comment]
Update: After looking at Milan's server logs from the crash, I believe the hang occurs during decode. In an attempt to simplify the repro, I am currently running a traced, decode-only 24-hour stress test on the following systems:
1. `sjc-snva-t3010`
2. `f10cs01`
3. `f10cs02`

All of these systems have been flashed to `80.13.1.0`.

**Comment by cglagovichTT** [comment]
Status:
`f10cs01`:
on `80.13.1.0`, decode trace test reproed the hang after 10 minutes. On a second run, it took 20m to repro the hang. 
on `80.13.1.1` (900MHz), decode trace test did not hang over the course of 12 hours. This indicates the hang is likely di/dt.

`f10cs02`:
on `80.13.1.0`, decode trace test reproed the hang after 4h25m.

`sjc-snva-t3010`:
on `80.13.1.0`, decode trace test has not yet reproed the hang after 18h19m.

Next steps:
- Leave decode trace test running on `sjc-snva-t3010` to see if it hangs.
- With `80.13.1.0` FW, run decode matmuls in a loop on `f10cs01` and `f10cs02` to see if the hang is isolated to one or more matmuls.

**Comment by cglagovichTT** [comment]
@ttmtrajkovic do you have a pointer to a test you've used to repro matmul di/dt? I want to make sure that if one of these matmuls is hanging then I'll catch it.

I'm tracing the matmul call and executing it in a loop with `blocking=True`, synchronizing devices every few iterations.

**Comment by ttmtrajkovic** [comment]
@cglagovichTT, the branch with didt tests is ppopovic/rebased_didt_tests.
The instructions to run tests are here: `tests/didt/README.md`.
There are there types of tests and LM head is the one where most of the failures are currently happening.

Milos

**Comment by cglagovichTT** [comment]
@ttmtrajkovic I isolated the failure to the LM head. it's a 32 core 1D matmul. On branch https://github.com/tenstorrent/tt-metal/tree/cglagovich/15090 repro with 
```
pytest  -svv models/demos/t3000/llama2_70b/tests/test_llama_didt_decode_matmul.py::test_decode_lm_head_matmul
```

Details of the matmul also pasted here:
```python
    memory_config = ttnn.MemoryConfig(
        ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        ttnn.BufferType.L1,
        ttnn.ShardSpec(
            num_to_corerange_set(32),
            [
                32,
                8192 // 32,
            ],
            ttnn.ShardOrientation.ROW_MAJOR,
            False,
        ),
    )
    xs = ttnn.as_tensor(
        torch.randn(1, 1, 32, 8192),  # seqlen=1, batch=1, batch_size=32, hidden_size
        device=mesh_device,
        dtype=ttnn.bfloat16,
        layout=ttnn.TILE_LAYOUT,
        memory_config=memory_config,
        mesh_mapper=ttnn.ReplicateTensorToMesh(mesh_device),
    )

    lm_head = ttnn.as_tensor(
        torch.randn(1, 1, 8192, 16 * 1024),  # hidden_size x padded_vocab_size
        device=mesh_device,
        dtype=ttnn.bfloat8_b,
        layout=ttnn.TILE_LAYOUT,
        memory_config=ttnn.DRAM_MEMORY_CONFIG,
        mesh_mapper=ttnn.ReplicateTensorToMesh(mesh_device),
    )

    program_config = ttnn.MatmulMultiCoreReuseMultiCast1DProgramConfig(
        compute_with_storage_grid_size=(8, 4),
        in0_block_w=8,
        out_subblock_h=1,
        out_subblock_w=4,
        per_core_M=1,
        per_core_N=16,
        fuse_batch=True,
        fused_activation=None,
        mcast_in0=True,
    )

    def run_lm_head():
        ttnn.matmul(
            xs,
            lm_head,
            program_config=program_config,
            memory_config=ttnn.DRAM_MEMORY_CONFIG,
            dtype=ttnn.bfloat16,
            compute_kernel_config=model_config["COMPUTE_KERNEL_CONFIG"],
        )

```

**Comment by cglagovichTT** [comment]
Update: 
We opened a side chat to discuss this issue. Over the last few days we've had a few findings.

I noticed that `f10cs01` dev0 and dev3 were operating at 75C during the hanging test, while other devices and `f10cs02` were cooler at 55C. `f10cs01` hung significantly sooner than `f10cs02`, so we suspected that thermals were one cause of hangs.
 Systems folks opened up `f10cs01` and noticed that the fans were installed in the wrong ports. Fixing that resolved the high temperatures I saw on `f10cs01`.

Beyond that, I'm reproing hangs in various ways.
- LM Head hang still exists
- 80 layer trace test (minus LM head) hangs
- 80 layer trace test (minus LM head and MLP) hangs
I see these hangs even with 1x1 subblocks on all matmuls. I am still trying to isolate the hangs to a specific op.

@pavlepopovic has not been able to repro the 1x4 subblock LM head hang on his t3k. We're looking into the possibility that these hangs are board issues and not di/dt.

**Comment by milank94** [comment]
Awaiting swap of cards with reworked card from: https://github.com/tenstorrent/cloud/issues/3448 to test if that resolves the issue.

**Comment by dthomp-ai** [comment]
@milank94 are you able to run any workloads on the systems in Serbia? There are 75+ reworked boards deployed and ready to go.

**Comment by milank94** [comment]
> [@milank94](https://github.com/milank94) are you able to run any workloads on the systems in Serbia? There are 75+ reworked boards deployed and ready to go.

Definitely. If we have a LoudBox configuration (4x n300) with reworked cards I can jump on that. Who should I reach out to for access?

**Comment by dthomp-ai** [comment]
> > [@milank94](https://github.com/milank94) are you able to run any workloads on the systems in Serbia? There are 75+ reworked boards deployed and ready to go.
> 
> Definitely. If we have a LoudBox configuration (4x n300) with reworked cards I can jump on that. Who should I reach out to for access?

I'll reach out to Zlatomir in Serbia and we'll track down someone on the Software team who knows. Stand by.

**Comment by milank94** [comment]
Update:

- Tested model workload for over 28 hours on reworked cards in Belgrade system with no issues. Will get more reworked samples next week in COLO to test.

**Comment by ttmtrajkovic** [comment]
Thanks @milank94, that's great news.

I think this issues should remain open with P2 priority until you are able to bring back identical reworked cards.
It's probably not on you to provide these cards, so it should be reassigned properly until reworked cards are available for this system to be back in.

Milos

**Comment by ttmtrajkovic** [comment]
@milank94,
The email thread I have started on establishing a unit test suite should be completed - the team has provided in the scattered form number of tests to cover: OPs, dispatch, ETH, CCL

I'd suggest to take this opportunity and create a suite for screening new systems, probably iterate with test owners few times, and then get this solid for all systems you have today and any future ones. 

That should be a much better practice than running Llama model right away. 


Milos

**Comment by milank94** [comment]
> Thanks [@milank94](https://github.com/milank94), that's great news.
> 
> I think this issues should remain open with P2 priority until you are able to bring back identical reworked cards. It's probably not on you to provide these cards, so it should be reassigned properly until reworked cards are available for this system to be back in.
> 
> Milos

Noted. I will lower priority and keep open for now. We are tracking the replacement of all of the other systems identified with this issue in our COLO here: https://github.com/tenstorrent/cloud/issues/3217. Cloud team is managing the swap.

**Comment by milank94** [comment]
> [@milank94](https://github.com/milank94), The email thread I have started on establishing a unit test suite should be completed - the team has provided in the scattered form number of tests to cover: OPs, dispatch, ETH, CCL
> 
> I'd suggest to take this opportunity and create a suite for screening new systems, probably iterate with test owners few times, and then get this solid for all systems you have today and any future ones.
> 
> That should be a much better practice than running Llama model right away.
> 
> Milos

Sounds good. I think we will open a separate tracker for this work. CC @mraumannTT

**Comment by prajaramanTT** [comment]
@milank94 Is this still an open issue ? If not. can you please close this ? Thanks.

**Comment by milank94** [comment]
Will close for now, but closely monitoring to see if these issue arise with reworked cards.

**Comment by ttmtrajkovic** [comment]
@milank94, did you get a chance to test this with exact cards that were reworked?

**Comment by milank94** [comment]
> [@milank94](https://github.com/milank94), did you get a chance to test this with exact cards that were reworked?

No. We would need to track down if and when those cards got reworked after they were removed from the original system.

@hmohiuddinTT who can help track down those original cards?

**Comment by milank94** [comment]
Closing since this is now a stale issue. We have not seen any issues reported with reworked cards in the field. Firmware releases have been updated since.
